File: android-lib/src/main/java/com/kinvey/android/async/AsyncDownloadRequest.java
package com.kinvey.android.async;


import com.kinvey.android.callback.AsyncDownloaderProgressListener;
import com.kinvey.java.Logger;
import com.kinvey.java.core.DownloaderProgressListener;
import com.kinvey.java.core.MediaHttpDownloader;

import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

public class AsyncDownloadRequest<T> extends AsyncRequest<T> {

    DownloaderProgressListener listener;

    public AsyncDownloadRequest(Object scope, Method method, final AsyncDownloaderProgressListener<T> callback, Object... args) {
        super(scope, method, callback, args);
        listener = new DownloaderProgressListener() {
            @Override
            public void progressChanged(final MediaHttpDownloader downloader) throws IOException {
                Runnable myRunnable = new Runnable() {
                    @Override
                    public void run() {
                        if (getCallback() != null) {
                            Logger.INFO("notifying callback");
                            try {
                                callback.progressChanged(downloader);
                            } catch (IOException e) {
                                callback.onFailure(e);
                            }
                        }

                    }
                };
                kinveyCallbackHandler.post(myRunnable);
            }
        };
    }

    @Override
    public T executeAsync() throws IOException, InvocationTargetException, IllegalAccessException {
        Object[] newArgs = new Object[args.length + 1];
        for (int i = 0; i < args.length; i++) {
            newArgs[i] = args[i];
        }
        newArgs[args.length] = listener;
        T ret = (T) mMethod.invoke(scope, newArgs);
        return ret;
    }

}

File: android-lib/src/main/java/com/kinvey/android/async/AsyncDownloadRequest.kt
package com.kinvey.android.async


import com.kinvey.android.callback.AsyncDownloaderProgressListener
import com.kinvey.java.Logger
import com.kinvey.java.core.DownloaderProgressListener
import com.kinvey.java.core.MediaHttpDownloader

import java.io.IOException
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method

class AsyncDownloadRequest<T>(scope: Any, method: Method?, callback: AsyncDownloaderProgressListener<T>?, vararg args: Any?) : AsyncRequest<T>(scope, method, callback, *args) {

    var listener: DownloaderProgressListener? = null

    init {
        listener = object : DownloaderProgressListener {
            @Throws(IOException::class)
            override fun progressChanged(downloader: MediaHttpDownloader?) {
                val myRunnable = Runnable {
                    if (callback != null) {
                        Logger.INFO("notifying callback")
                        try {
                            callback?.progressChanged(downloader)
                        } catch (e: IOException) {
                            callback?.onFailure(e)
                        }

                    }
                }
                kinveyCallbackHandler.post(myRunnable)
            }
        }
    }

    @Throws(IOException::class, InvocationTargetException::class, IllegalAccessException::class)
    override fun executeAsync(): T {
        val newArgs = arrayOf(*args, listener)
        return mMethod?.invoke(scope, *newArgs) as T
    }
}
